zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -118,6 +166,48 @@ impl Chain {
     pub fn is_empty(&self) -> bool {
         self.blocks.is_empty()
     }
+
+    pub fn history_root_hash(&self) -> ChainHistoryMmrRootHash {
+        self.history_tree.hash()
+    }
+
+    /// Clone the Chain but not the history tree, using the history tree
+    /// specified instead.
+    ///
+    /// Useful when forking, where the history tree is rebuilt anyway.
+    fn with_history_tree(&self, history_tree: HistoryTree) -> Self {
+        Chain {
+            blocks: self.blocks.clone(),
+            height_by_hash: self.height_by_hash.clone(),
+            tx_by_hash: self.tx_by_hash.clone(),
+            created_utxos: self.created_utxos.clone(),
+            spent_utxos: self.spent_utxos.clone(),
+            sprout_anchors: self.sprout_anchors.clone(),
+            sapling_anchors: self.sapling_anchors.clone(),
+            sprout_nullifiers: self.sprout_nullifiers.clone(),
+            sapling_nullifiers: self.sapling_nullifiers.clone(),
+            orchard_nullifiers: self.orchard_nullifiers.clone(),
+            partial_cumulative_work: self.partial_cumulative_work,
+            history_tree,
+        }
+    }
+
+    /// Return if two Chains are identical.
+    #[cfg(test)]
+    pub(crate) fn is_identical(&self, other: &Self) -> bool {
+        self.blocks == other.blocks
+            && self.height_by_hash == other.height_by_hash
+            && self.tx_by_hash == other.tx_by_hash
+            && self.created_utxos == other.created_utxos
+            && self.spent_utxos == other.spent_utxos
+            && self.sprout_anchors == other.sprout_anchors
+            && self.sapling_anchors == other.sapling_anchors
+            && self.sprout_nullifiers == other.sprout_nullifiers
+            && self.sapling_nullifiers == other.sapling_nullifiers
+            && self.orchard_nullifiers == other.orchard_nullifiers
+            && self.partial_cumulative_work == other.partial_cumulative_work
+            && self.history_root_hash() == other.history_root_hash()
+    }
 }
```

Comment:
Is this different than `Eq`?

Reply:
> Is this different than `Eq`?

`self.history_root_hash() == other.history_root_hash()` won't be part of a derived `PartialEq` on `Chain`.

But we could fix that by implementing `PartialEq` and `Eq` on `HistoryTree`.

Reply:
I've implemented `PartialEq` and `Eq` on `HistoryTree` to makes `Chain`s comparable with `derive(Eq)`.

However I had to remove the existing `Eq` implementation of `Chain` that used `cmp`. I think that's a not an issue because the existing `cmp` never returns equal anyway (it assumes it will never compare equal chains), but please double check.

(I've cheated a little bit and kept using `proptest_assert` (instead of `_eq`) to compare `Chain`s because that seems to require `Debug` and I can't  `derive` it because some underlying `librustzcash` types do not implement it, and I wanted to get this done today. But I can implement `Debug` manually later.)

Done in 8dae6e73

Reply:
Added `Debug` implementations in 2e928db9